enhancements/dns/aro-private-dns-zone-resource-removal.md

```diff
@@ -0,0 +1,180 @@
+
+1. ARO intends to enable zero Internet egress by default for ARO clusters.  It
+   is likely that in order to implement this, ARO will need to be able to
+   selectively override DNS resolution of certain domains in both node and pod
+   contexts.
+
+### Goals
+
+* ARO cluster functionality should not be dependent on Azure vnet DNS
+  configuration.  Azure private DNS zone resources should not be part of the ARO
+  architecture.  The changes required are implemented in ARO, not OCP.
+
+### Non-Goals
+
+## Proposal
+
+ARO would like OCP's approval to remove the private DNS zone resource from its
+cluster architecture.  Instead, we propose to resolve
+{api,api-int,\*.apps}.*cluster domain* DNS addresses via an inline dnsmasq proxy
+resolver running as a systemd service on each cluster VM.  The dnsmasq proxy
+resolver would be backed by whichever DNS server(s) the customer specifies in
+their vnet configuration.
+
+A working PoC-quality implementation of this approach can be found at
+https://github.com/Azure/ARO-RP/pull/1296.  This implementation is notably
+similar to the following [Machine Config Operator
+code](https://github.com/openshift/machine-config-operator/blob/master/templates/common/on-prem/files/coredns-corefile.yaml),
+however by using dnsmasq which is already present in the RHCOS image, it does
+not depend on the ability to pull container images.  This will be valuable when
+implementing zero Internet egress.
+
+We do not believe that the PoC implementation is trivially applicable to the
+OpenShift installer because it is dependent on calculating the IP addresses for
+{api,api-int,\*.apps}.*cluster domain* before the installer graph is created.
+We suppose this might imply larger installer architecture changes in the
+OpenShift installer IPI general case were this ever to be implemented there, but
+it is relatively straightforward in ARO.
+
+We do believe that the architectural changes/constraints placed on OCP could be
+feasible.  These are:
+
+* ARO clusters would no longer have associated Azure private DNS zones.
+
+* Dnsmasq would be used as an inline proxying DNS resolver for all ARO cluster
+  node-context and pod-context DNS resolution.
+
+* Dnsmasq would need to continue to be provided in the RHCOS image.
+
+* ARO VMs would no longer be DNS resolvable by their hostname from other VMs.
+  Note that because --kubelet-preferred-address-types is set to InternalIP in
+  the Kubelet config, we don't believe this is currently a hard platform
+  requirement.
+
+* ARO machine config server URL and serving certificate would use api-int IP
+  address instead of DNS.  This is because DNS resolution of api-int would not
+  be available at ignition time.
+
+* The ARO cluster installation process would prepopulate the
+  openshift-ingress/router-default LoadBalancer service early with the
+  appropriate IP address.
+
+* Ingress operator dynamic DNS would be disabled on ARO via overridden
+  configuration of the DNS custom resource.
```

Comment:
Do you mean the operator's publishing to the private zone would be disabled by overriding the private zone in the cluster DNS config, or are you trying to disable all DNS publishing by the operator?

Reply:
Yes. I managed services customer does not own or manage DNS and cluster does not have any access to do so. So it is managed from outside. This should not be for scope for everybody but we need this for ARO specifically.